File: projects/browser-recorder/browser_recorder/inject.py
"""事件捕获注入：RECORDER_JS + install(context)。

设计要点：
- context 级 add_init_script + expose_function，一次安装，所有导航/新 tab 自动生效
  （CDP attach 已有页面时对"未来导航"生效，调用方需 goto/reload 强制注入）。
- JS 侧算好 SelectorSet 全候选（testid → role+name → id → text → css）。
- input 防抖：change/blur 定稿，或 500ms 静默定稿。
- 防丢：payload 先进 sessionStorage.__br_pending，导航后重跑时 flush 补发（id 去重）。
- iframe：window.__recordEvent 不存在时尝试 parent，跨域 catch 后 console.warn 丢弃。
"""

from __future__ import annotations

from typing import Callable

RECORDER_JS = r"""
(() => {
  if (window.__brInstalled) return;
  window.__brInstalled = true;

  const PENDING_KEY = "__br_pending";
  let seq = 0;

  function bridge() {
    if (typeof window.__recordEvent === "function") return window.__recordEvent;
    try {
      if (window.parent && typeof window.parent.__recordEvent === "function") {
        return window.parent.__recordEvent;
      }
    } catch (e) { /* 跨域 iframe */ }
    return null;
  }

  function enqueue(payload) {
    payload.id = `${Date.now()}-${++seq}`;
    try {
      const pending = JSON.parse(sessionStorage.getItem(PENDING_KEY) || "[]");
      pending.push(payload);
      sessionStorage.setItem(PENDING_KEY, JSON.stringify(pending.slice(-200)));
    } catch (e) { /* sessionStorage 不可用时直接发 */ }
    deliver(payload);
  }

  function deliver(payload) {
    const fn = bridge();
    if (fn) {
      try { fn(JSON.stringify(payload)); } catch (e) { /* 页面销毁中 */ }
    } else {
      console.warn("[browser-recorder] 无 __recordEvent 桥，事件暂存: " + payload.type);
    }
  }

  // 导航后重跑时补发暂存事件（幂等靠 Python 侧 payload.id 去重）
  function flushPending() {
    if (typeof window.__recordEvent !== "function") return;
    let pending;
    try {
      pending = JSON.parse(sessionStorage.getItem(PENDING_KEY) || "[]");
      sessionStorage.removeItem(PENDING_KEY);
    } catch (e) { return; }
    for (const p of pending) deliver(p);
  }

  // ---------- SelectorSet 生成 ----------

  function isDynamicIdent(s) {
    // ember123 / css-1x2y3z / :r0: 这类动态 id/class 不稳定，不采用
    return /(^|[-_:])\d{2,}/.test(s) || /[0-9a-f]{8,}/i.test(s);
  }

  function cssEscape(s) {
    if (window.CSS && CSS.escape) return CSS.escape(s);
    return s.replace(/([^\w-])/g, "\\$1");
  }

  function textOf(el) {
    const t = (el.innerText || el.value || el.getAttribute("aria-label") || "").trim();
    return t.length > 40 ? t.slice(0, 40) : t;
  }

  function roleOf(el) {
    const explicit = el.getAttribute("role");
    if (explicit) return explicit;
    const tag = el.tagName.toLowerCase();
    const type = (el.getAttribute("type") || "").toLowerCase();
    if (tag === "button" || (tag === "input" && ["button", "submit", "reset"].includes(type))) return "button";
    if (tag === "a" && el.href) return "link";
    if (tag === "select") return "combobox";
    if (tag === "textarea" || (tag === "input" && !["checkbox", "radio"].includes(type))) return "textbox";
    if (tag === "input" && type === "checkbox") return "checkbox";
    if (tag === "input" && type === "radio") return "radio";
    return null;
  }

  function labelOf(el) {
    // 关联 label > aria-label > placeholder > 元素自身文本
    if (el.id) {
      const lbl = document.querySelector(`label[for="${cssEscape(el.id)}"]`);
      if (lbl && lbl.innerText.trim()) return lbl.innerText.trim().slice(0, 40);
    }
    return (el.getAttribute("aria-label") || el.getAttribute("placeholder") || textOf(el) || el.name || null);
  }

  function cssPath(el) {
    const parts = [];
    let cur = el;
    while (cur && cur.nodeType === 1 && cur !== document.body) {
      let part = cur.tagName.toLowerCase();
      if (cur.id && !isDynamicIdent(cur.id)) {
        part += "#" + cssEscape(cur.id);
        parts.unshift(part);
        break;
      }
      const stable = [...cur.classList].filter(c => !isDynamicIdent(c)).slice(0, 2);
      if (stable.length) part += "." + stable.map(cssEscape).join(".");
      const parent = cur.parentElement;
      if (parent) {
        const same = [...parent.children].filter(c => c.tagName === cur.tagName);
        if (same.length > 1) part += `:nth-of-type(${same.indexOf(cur) + 1})`;
      }
      parts.unshift(part);
      cur = parent;
    }
    return parts.join(" > ");
  }

  function selectorsFor(el) {
    const out = {};
    const testid = el.getAttribute("data-testid") || el.getAttribute("data-test-id");
    if (testid) out.testid = `[data-testid="${testid}"]`;
    const role = roleOf(el);
    const label = labelOf(el);
    if (role && label) out.role_name = `role=${role}[name="${label.replace(/"/g, '\\"')}"]`;
    else if (role) out.role_name = `role=${role}`;
    if (el.id && !isDynamicIdent(el.id)) out.id = "#" + cssEscape(el.id);
    const text = textOf(el);
    if (text && ["button", "link"].includes(role)) out.text = `text=${text}`;
    out.css = cssPath(el);
    return out;
  }

  // ---------- 事件捕获 ----------

  function basePayload(type, el) {
    return {
      type,
      url: location.href,
      selectors: el ? selectorsFor(el) : {},
      label: el ? labelOf(el) : null,
    };
  }

  document.addEventListener("click", (e) => {
    const el = e.target.closest("button, a, input[type=submit], input[type=button], [role=button], select, summary")
      || e.target;
    const p = basePayload("click", el);
    p.point = { x: Math.round(e.clientX), y: Math.round(e.clientY) };
    enqueue(p);
  }, true);

  // mousedown 提前通知：click 的默认行为/页面响应在 mousedown 之后才发生，
  // 此时截图是"点击前"的干净画面。通过独立的 __recordMouseDown 桥立即通知
  // Python 截图（同步阶段触发，赶在页面响应前）。
  document.addEventListener("mousedown", (e) => {
    const fn = window.__recordMouseDown ||
      (() => { try { return window.parent && window.parent.__recordMouseDown; } catch (err) { return null; } })();
    if (typeof fn !== "function") return;
    try {
      fn(JSON.stringify({ x: Math.round(e.clientX), y: Math.round(e.clientY), url: location.href }));
    } catch (err) { /* 忽略 */ }
  }, true);

  // input 防抖：静默 500ms 或 change/blur 定稿
  let pendingInput = null;
  let inputTimer = null;

  function commitInput() {
    if (pendingInput) {
      enqueue(pendingInput);
      pendingInput = null;
    }
    if (inputTimer) { clearTimeout(inputTimer); inputTimer = null; }
  }

  document.addEventListener("input", (e) => {
    const el = e.target;
    if (!el.matches("input, textarea")) return;
    const type = (el.getAttribute("type") || "text").toLowerCase();
    if (["checkbox", "radio", "button", "submit", "file"].includes(type)) return;
    const p = basePayload("input", el);
    p.value = el.value;
    p.sensitive = type === "password";
    pendingInput = p;
    if (inputTimer) clearTimeout(inputTimer);
    inputTimer = setTimeout(commitInput, 500);
  }, true);

  document.addEventListener("change", (e) => {
    const el = e.target;
    if (el.tagName === "SELECT") {
      commitInput();
      const p = basePayload("select", el);
      p.value = el.value;
      p.label = labelOf(el);
      enqueue(p);
    } else if (el.matches("input, textarea")) {
      commitInput(); // input 已有 pending 则定稿（change 拿到的值相同）
    }
  }, true);

  document.addEventListener("blur", commitInput, true);

  document.addEventListener("keydown", (e) => {
    if (e.key === "Enter" || e.key === "Tab") {
      commitInput();
      if (e.key === "Enter") {
        const p = basePayload("key", e.target);
        p.value = "Enter";
        enqueue(p);
      }
    }
  }, true);

  // SPA 路由变化：hook history API，记 navigate 步骤（无整页导航时）
  let lastUrl = location.href;
  function onRouteChange() {
    if (location.href === lastUrl) return;
    lastUrl = location.href;
    const p = basePayload("navigate", null);
    p.value = location.href;
    enqueue(p);
  }
  for (const m of ["pushState", "replaceState"]) {
    const orig = history[m];
    history[m] = function (...args) {
      const r = orig.apply(this, args);
      onRouteChange();
      return r;
    };
  }
  window.addEventListener("popstate", onRouteChange);
  window.addEventListener("hashchange", onRouteChange);

  // 初次注入（含每次导航后）补发暂存事件
  if (document.readyState === "loading") {
    document.addEventListener("DOMContentLoaded", flushPending);
  } else {
    flushPending();
  }
})();
"""


def install(context, on_event: Callable[[dict], None], on_mousedown: Callable[[dict], None] | None = None) -> None:
    """在 context 安装事件捕获：注入脚本 + 暴露 __recordEvent 桥。

    重要：binding 用 **page 级** expose_function，不用 context 级。
    实测（playwright 1.62 + CDP attach）：context 级 expose_function 注册的
    binding 在 attach 的 context 上不触发回调（init script 正常，binding 消息
    不回本 session）；page 级正常。因此 install 时对已有 page 逐个暴露，
    并通过 context.on("page") 给未来新 tab 补挂。
    init script 仍用 context 级（对新导航/新 tab 自动生效，实测正常）。

    CDP attach 已有页面时，调用方需在 install 后 goto/reload 一次保证注入。
    """
    import json

    def _make_bridge():
        def _bridge(payload: str) -> None:
            try:
                on_event(json.loads(payload))
            except Exception:
                pass  # 单个事件解析失败不中断录制
        return _bridge

    def _make_mousedown_bridge():
        def _bridge(payload: str) -> None:
            try:
                if on_mousedown:
                    on_mousedown(json.loads(payload))
            except Exception:
                pass
        return _bridge

    def _expose(page) -> None:
        try:
            page.expose_function("__recordEvent", _make_bridge())
        except Exception:
            pass  # 页面关闭中等场景忽略
        try:
            page.expose_function("__recordMouseDown", _make_mousedown_bridge())
        except Exception:
            pass

    for page in context.pages:
        _expose(page)
    context.on("page", _expose)
    context.add_init_script(RECORDER_JS)

File: projects/browser-recorder/src/browser_recorder/inject.js
/* browser-recorder 页面注入脚本。
 * 两种形态：
 *  - node（require）：导出纯函数（best_selector/trunc_text）供单测
 *  - 浏览器（CDP 注入字符串）：IIFE install，监听动作/突变/停止热键，经 __brEvent 上报
 *
 * payload 为扁平结构 {type, rect, viewport, descriptor, value?, html_type?}，
 * 与 recorder.py 的 Runtime.bindingCalled 消费端约定一致。
 */
(function (root, factory) {
  if (typeof module === "object" && module.exports) { module.exports = factory(); }
  else { factory().install(root); }
})(typeof self !== "undefined" ? self : this, function () {
  "use strict";

  var TEXT_MAX = 40;
  var VALUE_MAX = 200;
  var DOM_PATH_MAX = 12;
  var MUTATION_WINDOW_MS = 150;

  function trunc_text(s) {
    s = String(s == null ? "" : s).replace(/\s+/g, " ").trim();
    if (s.length > TEXT_MAX) s = s.slice(0, TEXT_MAX - 1) + "…";
    return s;
  }

  function best_selector(desc) {
    if (desc.id) return "#" + desc.id;
    var parts = [desc.tag];
    if (desc.classes && desc.classes.length) parts.push("." + desc.classes.join("."));
    return parts.join("");
  }

  function classes_of(el) {
    // SVG/HTML 均有 classList（DOMTokenList）；个别旧实现缺省时退回 className 解析
    try {
      if (el.classList) return Array.prototype.slice.call(el.classList);
    } catch (e) { /* 某些元素访问 classList 可能抛 */ }
    var cn = el.className;
    if (typeof cn !== "string") cn = (cn && cn.baseVal) || ""; // SVG 的 className 是 SVGAnimatedString
    return cn.split(/\s+/).filter(Boolean);
  }

  function dom_path(el) {
    // parentElement 在 closed shadow root 边界处为 null，链自然截断（不跨 shadow）
    var p = [], cur = el;
    while (cur && cur.nodeType === 1 && p.length < DOM_PATH_MAX) {
      p.unshift(cur.tagName.toLowerCase() + (cur.id ? "#" + cur.id : ""));
      cur = cur.parentElement;
    }
    return p.join(">");
  }

  function describe(el, win) {
    win = win || (typeof self !== "undefined" ? self : globalThis);
    var r = el.getBoundingClientRect ? el.getBoundingClientRect() : { x: 0, y: 0, width: 0, height: 0 };
    var desc = {
      tag: (el.tagName || "").toLowerCase(),
      id: el.id || null,
      classes: classes_of(el),
      text: trunc_text(
        el.textContent ||
          (el.type === "password" ? "" : el.value) ||
          (el.getAttribute && el.getAttribute("aria-label")) ||
          ""
      ),
      dom_path: dom_path(el),
    };
    desc.best_selector = best_selector(desc);
    return {
      rect: {
        x: Math.round(r.x || 0), y: Math.round(r.y || 0),
        w: Math.round(r.width || 0), h: Math.round(r.height || 0),
      },
      viewport: {
        w: win.innerWidth || 0, h: win.innerHeight || 0,
        scrollX: win.scrollX || 0, scrollY: win.scrollY || 0,
        dpr: win.devicePixelRatio || 1,
      },
      descriptor: desc,
    };
  }

  function install(win) {
    var B = win.__brEvent;
    if (!B || win.__brInstalled) return; // binding 未就绪时不设哨兵，允许下次注入重试
    win.__brInstalled = true;

    function report(type, el, extra) {
      try {
        var payload = describe(el, win);
        payload.type = type;
        if (extra) for (var k in extra) payload[k] = extra[k];
        B(JSON.stringify(payload));
      } catch (e) { /* 上报失败不扰动页面 */ }
    }

    function target_of(e) {
      // composedPath 穿透 shadow：open shadow 内的事件 target 在主文档侧
      // 表现为 shadow host（如 EO-LAUNCHPAD-BUTTON-V2），真实交互元素只在
      // composedPath 里可见。取路径上第一个可交互元素。
      var path = e.composedPath ? e.composedPath() : [e.target];
      var interact = "a,button,input,select,textarea,[onclick],[role=button],[contenteditable]";
      for (var i = 0; i < path.length; i++) {
        var node = path[i];
        if (node && node.closest) {
          var hit = node.closest ? (node.closest(interact) || null) : null;
          if (hit) return hit;
          if (node.matches && node.matches(interact)) return node;
        }
      }
      var t = e.target;
      if (t && t.closest) return t.closest(interact) || t;
      return t;
    }

    // input 击键聚合：同一输入框的连续击键（<1.2s 间隔）不逐键上报——上报
    // 会逐键截图 + 逐键落 action（用户录 easyops 六个字符产生 8 个动作噪声）。
    // 收尾（blur/换目标/超时）时 flush 最终值。
    var lastInput = null; // {el, timer}
    var INPUT_FLUSH_MS = 1200;
    function flushInput() {
      if (!lastInput) return;
      var li = lastInput; lastInput = null;
      clearTimeout(li.timer);
      try {
        var type = li.el.type || "text";
        report("input", li.el, {
          value: type === "password" ? "***" : String(li.el.value == null ? "" : li.el.value).slice(0, VALUE_MAX),
          html_type: type,
          composed: true,
        });
      } catch (e) { /* ignore */ }
    }

    // 动作：capture 阶段，document 级。click/submit/change 都冲刷在途输入
    // 聚合（真实操作里输入总是被下一步动作或 change 收尾）。
    ["click", "submit", "change"].forEach(function (t) {
      win.document.addEventListener(t, function (e) {
        if (t !== "submit" || e.target && e.target.form !== undefined) flushInput();
        var el = t === "submit" ? e.target : target_of(e);
        if (t === "change") return; // change 只用于冲刷，不产生动作事件
        if (el) report(t, el, {});
      }, true);
    });

    win.document.addEventListener("input", function (e) {
      var el = e.target;
      if (!el) return;
      if (lastInput && lastInput.el === el) {
        // 同一输入框：重置计时（击键间隔刷新），不立即上报
        clearTimeout(lastInput.timer);
        lastInput.timer = setTimeout(flushInput, INPUT_FLUSH_MS);
        return;
      }
      flushInput(); // 换了输入框，先结上一个
      lastInput = { el: el, timer: setTimeout(flushInput, INPUT_FLUSH_MS) };
    }, true);

    win.document.addEventListener("beforeunload", function () { flushInput(); }, true);
    // 录制器停止前主动调用（Browser.close 不走页面 unload，挂起中的输入
    // 事件会随进程消亡丢失）：window.__brFlush()
    win.__brFlush = flushInput;

    // 停止热键：Ctrl+Shift+F9（key 或 keyCode 120），覆盖所有 frame 的注入实例
    win.document.addEventListener("keydown", function (e) {
      if (e.ctrlKey && e.shiftKey && (e.key === "F9" || e.keyCode === 120)) {
        e.preventDefault();
        e.stopPropagation();
        try { B(JSON.stringify({ type: "control_stop" })); } catch (err) { /* ignore */ }
      }
    }, true);

    // DOM 突变聚合：150ms 静默窗口。
    // 注入可能发生在新文档极早期（Page.addScriptToEvaluateOnNewDocument 在
    // documentElement 创建前运行），此时 observe(null) 会抛 TypeError 导致
    // install 半途夭折—— MutationObserver 是 install 最后一步，动作监听器
    // 已先挂上，症状是"动作有、dom_mutations 永远没有"。故延迟到根元素可用。
    var pending = 0;
    var timer = null;
    var obs = new MutationObserver(function (muts) {
      pending += muts.length;
      clearTimeout(timer);
      timer = setTimeout(function () {
        var n = pending;
        pending = 0;
        try { B(JSON.stringify({ type: "dom_mutations", count: n })); } catch (e) { /* ignore */ }
      }, MUTATION_WINDOW_MS);
    });
    function startObserver() {
      try {
        obs.observe(win.document.documentElement, {
          subtree: true, childList: true, attributes: true, characterData: true,
        });
        return true;
      } catch (e) { return false; }
    }
    if (!startObserver()) {
      var pollT = setInterval(function () {
        if (win.document.documentElement && startObserver()) clearInterval(pollT);
      }, 10);
    }
  }

  return {
    best_selector: best_selector,
    trunc_text: trunc_text,
    install: install,
    _describeForTest: describe,
  };
});
